### PR TITLE
Allow to list the rasters of a folder

### PR DESCRIPTION
**Pull request by ao-picterra**

### CODE REVIEW

File: src/picterra/client.py [hunk 1]
```diff
@@ -82,22 +82,29 @@ def _wait_until_operation_completes(self, operation_response):
...
+            params['page_number'] = page
+            logger.debug('Paginating through %s list at page %s' % (resource_endpoint, page))
+            resp = self.sess.get(url, params=params)
             if not resp.ok:
                 raise APIError(resp.text)
             r = resp.json()
-            url = r['next']
             count = r['count']
             data += r['results']
+            if not r['next']:
```
Comment (julienr) [inline src/picterra/client.py:98]: Is the change from `while url` do `while True` with the `if` fixing a bug of some kind ?

**Comment by julienr** [comment]
Thanks !

File: src/picterra/client.py (hunk 1 above)
Comment (restaste) [reply src/picterra/client.py:99]: If that's not necessary I'd maybe revert, because I kinda feel we should prefer using the `next` url rather than computing it ourselves (and having to keep track of the current page).

File: tests/test_client.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def api_url(path):
     return urljoin(TEST_API_URL, path)
 
 
-def add_mock_rasters_list_response():
+def add_mock_rasters_list_response(folder_id=''):
```
Comment (restaste) [inline tests/test_client.py:22]: I'd consider doing an extra `add_mock_rasters_in_folder_list_response` function to keep the level of branching a bit lower.

File: tests/test_module_cli.py [hunk 3]
```diff
@@ -19,17 +23,20 @@ def test_parser_arg_errors(capsys):
...
-    monkeypatch.setattr(APIClient, '__init__', lambda s: None)
+    monkeypatch.setattr(APIClient, '__init__', _fake__init__)
     mock_rasterlist = MagicMock(return_value=['foo', 'bar'])
     mock_rasterlist.called is False
     monkeypatch.setattr(APIClient, 'list_rasters', mock_rasterlist)
     parse_args(['list', 'rasters'])
     mock_rasterlist.called is True
-
+    mock_rasterlist.reset_mock()
+    mock_rasterlist.called is False
```
Comment (restaste) [inline tests/test_module_cli.py:33]: What does that line do? Actually, all of the `mock_rasterlist.called is ..`, do they work without `assert` ?

**Review by restaste — COMMENTED** [review]
Works well! Just a couple of details.

File: src/picterra/client.py (hunk 1 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked

File: tests/test_client.py (hunk 2 above)
Comment (ao-picterra) [reply tests/test_client.py]: Checked

File: tests/test_module_cli.py (hunk 3 above)
Comment (ao-picterra) [reply tests/test_module_cli.py]: Checked

File: src/picterra/client.py [hunk 4]
```diff
@@ -146,8 +158,13 @@ def upload_raster(self, filename, name, folder_id=None, captured_at=None):
         self._wait_until_operation_completes(resp.json())
         return raster_id
 
-    def list_rasters(self):
+    def list_rasters(self, folder_id: str = ''):
```
Comment (julienr) [inline src/picterra/client.py:161]: In Python, it's better to use None to indicate a parameter is not set. So I'd have `folder_id=None` here to indicate we don't want filtering

File: src/picterra/__main__.py [hunk 5]
```diff
@@ -19,8 +19,19 @@
...
 def parse_args(args):
+    # Instantiate client, just for checking URl and api key
+    try:
+        client = APIClient()
+        target = client.base_url
+        suffix = "Your API key starts with '%s'" % client.api_key[:4]
+        desc = "You are targeting %s.\n\n%s" % (target, suffix)
+        del client
+    except APIError:
+        desc = "ATTENTION: you do not have an Api Key: set one before using the cli commands."
```
Comment (julienr) [inline src/picterra/__main__.py:30]: This message won't be printed anywhere and we'll actually not show the underlying APIError. I wouldn't try to be too smart here and just let it fail

File: src/picterra/__main__.py [hunk 6]
```diff
@@ -19,8 +19,19 @@
 
 
 def parse_args(args):
+    # Instantiate client, just for checking URl and api key
+    try:
+        client = APIClient()
+        target = client.base_url
+        suffix = "Your API key starts with '%s'" % client.api_key[:4]
+        desc = "You are targeting %s.\n\n%s" % (target, suffix)
+        del client
```
Comment (julienr) [inline src/picterra/__main__.py:28]: I'd put this code into a `check_api_key` helper function, this removes the need for the `del`

File: examples/raster_management.py [hunk 7]
```diff
@@ -1,15 +1,27 @@
 #!/usr/bin/env python3
 
+from pprint import pprint
 from picterra import APIClient
 
 # Set the PICTERRA_API_KEY environment variable to define your API key
 client = APIClient()
+# The Id of a folder/project you own
+folder_id = "abcdefghijklmnopwrstuvz"
```
Comment (julienr) [inline examples/raster_management.py:9]: Let's show a proper uuid here, just to server as a good example of what the id should look like

File: src/picterra/client.py [hunk 8]
```diff
@@ -82,22 +83,38 @@ def _wait_until_operation_completes(self, operation_response):
...
         url = self._api_url('%s/?page_number=1' % resource_endpoint)
+        # Store list data, page (for debugging, future limit), total count (same as latter)
         data = []
+        page = 0
+        count = 0
+        # Loop over all the pages
         while url:
-            logger.debug('Paginating through %s list at page %s' % (resource_endpoint, url))
-            resp = self.sess.get(url)
+            page += 1
```
Comment (julienr) [inline src/picterra/client.py:95]: I feel this whole thing could be rewritten into:

```
data = []
url = self._api_url('%s/?page_number=1' % resource_endpoint, params=params)
while url:
  logger.debug('fetching page url=%s', url)
  resp = self.sess.get(url)
  if not resp.ok:
    raise APIError(resp.text)
  r = resp.json()
  url = r['next']
  count = r['count']
  data += r['results']
assert len(data) == count, 'Error in pagination [%d!=%d]' % (len(data), count)
return data
```

No need for the page/count bookkeeping. It should be included in the logged url anyway.

File: src/picterra/client.py [hunk 9]
```diff
@@ -106,23 +123,18 @@ def upload_raster(self, filename, name, folder_id=None, captured_at=None):
...
-        if folder_id is not None:
-            data.update({
-                'folder_id': folder_id
-            })
-        if captured_at is not None:
-            data.update({
-                'captured_at': captured_at
-            })
+        data = {'name': name}
+        if folder_id != '':
```
Comment (julienr) [inline src/picterra/client.py:134]: Let's use None, its the right thing to do in Python to indicate a param is not set. So I would revert this to what we had before

File: src/picterra/client.py [hunk 10]
```diff
@@ -343,7 +363,7 @@ def edit_detector(
         # Call API and check response
         resp = self.sess.put(
             self._api_url('detectors/%s/' % detector_id),
-            data=body_data
+            json=body_data
```
Comment (julienr) [inline src/picterra/client.py:366]: This seems unrelated and I think it's equivalent as far as requests is concerned. I'd chop this change

**Review by julienr — CHANGES_REQUESTED** [review]

File: src/picterra/client.py (hunk 1 above)
Comment (julienr) [reply src/picterra/client.py]: I added another comment, but I basically wouldn't add bookkeeping (it's always error prone) that we just use us when debugging the library. This is stuff you can add when debugging and remove afterwards. Let's see if one day we want to limit pagination - then we can add this, but for now I'd just keep it simple

File: src/picterra/client.py [hunk 11]
```diff
@@ -82,22 +83,38 @@ def _wait_until_operation_completes(self, operation_response):
                 raise APIError('Operation %s failed' % operation_id)
             time.sleep(poll_interval)
 
-    def _paginate_through_list(self, resource_endpoint: str):
+    def _paginate_through_list(self, resource_endpoint: str, params: dict = {}):
```
Comment (julienr) [inline src/picterra/client.py:86]: Don't use [mutable as default values](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html) in Python. The right way to write this is:
```
def _paginate_through_list(self, ..., params = None):
  if params is None:
    params = {}
```

File: src/picterra/client.py [hunk 12]
```diff
@@ -82,22 +83,38 @@ def _wait_until_operation_completes(self, operation_response):
...
             url = r['next']
             count = r['count']
             data += r['results']
-        assert len(data) == count
+        assert len(data) == count, 'Error in pagination [%d!=%d]' % (len(data), count)
         return data
 
-    def upload_raster(self, filename, name, folder_id=None, captured_at=None):
+    def upload_raster(
+        self, filename: str, name: str, folder_id: str = '', captured_at: str = ''
```
Comment (julienr) [inline src/picterra/client.py:116]: Use `None` instead of empty strings for unset parameters.

File: src/picterra/__main__.py [hunk 13]
```diff
@@ -19,8 +19,19 @@
 
 
 def parse_args(args):
+    # Instantiate client, just for checking URl and api key
+    try:
+        client = APIClient()
```
Comment (julienr) [inline src/picterra/__main__.py:24]: Wouldn't it be cleaner to just add an 'apikey' option to the cli that would print the api key ?

I feel we are a bit reinventing the wheel given your api key is in `PICTERRA_API_KEY`. Also, keep in mind users aren't like us juggling with test and production API keys - they have one so its unlikely it will be setup wrong, no ? So yeah not sure this whole thing is worth it to be honest. What is it solving ?

File: examples/raster_management.py (hunk 7 above)
Comment (ao-picterra) [reply examples/raster_management.py]: Checked

File: src/picterra/__main__.py (hunk 13 above)
Comment (ao-picterra) [reply src/picterra/__main__.py]: Checked (gor rid of all)

File: src/picterra/__main__.py (hunk 5 above)
Comment (ao-picterra) [reply src/picterra/__main__.py]: Checked (got rid of all)

File: src/picterra/client.py (hunk 11 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked

File: src/picterra/client.py (hunk 8 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked (I put your version, just modified `_api_url` to add querystring params

File: src/picterra/client.py (hunk 12 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked

File: src/picterra/client.py (hunk 9 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked

File: src/picterra/client.py (hunk 4 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked

File: src/picterra/client.py (hunk 10 above)
Comment (ao-picterra) [reply src/picterra/client.py]: Checked (reverted)

**Review by julienr — APPROVED** [review]

**Comment by julienr** [comment]
Thanks !
